iree/hal/local/sync_semaphore.h

```diff
@@ -0,0 +1,75 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IREE_HAL_LOCAL_SYNC_SEMAPHORE_H_
+#define IREE_HAL_LOCAL_SYNC_SEMAPHORE_H_
+
+#include "iree/base/api.h"
+#include "iree/base/synchronization.h"
+#include "iree/hal/api.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+//===----------------------------------------------------------------------===//
+// iree_hal_sync_semaphore_state_t
+//===----------------------------------------------------------------------===//
+
+// State shared between all sync semaphores.
+// Owned by the device and guaranteed to remain valid for the lifetime of any
+// semaphore created from it.
+typedef struct {
+  // In-process notification signaled when any semaphore value changes.
+  iree_notification_t notification;
+} iree_hal_sync_semaphore_state_t;
+
+// Initializes state used to perform semaphore synchronization.
+void iree_hal_sync_semaphore_state_initialize(
+    iree_hal_sync_semaphore_state_t* out_shared_state);
+
+// Deinitializes state used to perform semaphore synchronization; no semaphores
+// must be live with references.
+void iree_hal_sync_semaphore_state_deinitialize(
+    iree_hal_sync_semaphore_state_t* shared_state);
+
+//===----------------------------------------------------------------------===//
+// iree_hal_sync_semaphore_t
+//===----------------------------------------------------------------------===//
+
+// Creates a semaphore that allows for ordering of operations on the local host.
+iree_status_t iree_hal_sync_semaphore_create(
+    iree_hal_sync_semaphore_state_t* shared_state, uint64_t initial_value,
+    iree_allocator_t host_allocator, iree_hal_semaphore_t** out_semaphore);
+
+// Performs a signal of a list of semaphores.
+// The semaphores will transition to their new values (nearly) atomically and
+// batching up signals will reduce synchronization overhead.
+iree_status_t iree_hal_sync_semaphore_multi_signal(
+    iree_hal_sync_semaphore_state_t* shared_state,
+    const iree_hal_semaphore_list_t* semaphore_list);
```

Comment:
Are these semaphores assumed to be created by `iree_hal_sync_semaphore_create`? looks like the `iree_hal_semaphore_t` type is reused for each implementation. Can implementations be mixed in a list?

Reply:
yeah not legal to mix HAL types from other devices (we'd need explicit import/export APIs for that), so safe to assume this (some places have some type guards but more could be added whenever we allow multiple simultaneous HAL usage - today there's only ever one HAL device at a time so not something that can be hit)